Port to Python 3

exarkun opened the pull request:
Fixes #276 

This drops Python 2 support.

tomprince commented on src/_zkapauthorizer/tests/test_plugin.py:
```diff
@@ -625,25 +614,64 @@ def _create(self, get_config, servers_yaml, rootcap):
+        # and unicode in an os.path.join() call that always fails with a
+        # TypeError.
+        def write_private_config(name, value):
+            privname = os.path.join(config._basedir, u"private", name)
+            with open(privname, "wb") as f:
+                f.write(value)
```
`FilePath` or `pathlib`?

tomprince commented on src/_zkapauthorizer/tests/test_pricecalculator.py:
```diff
@@ -21,6 +21,7 @@
 
 from hypothesis import given
 from hypothesis.strategies import integers, lists, tuples
+from past.builtins import long
```
Can we do `s/long/int` instead?

tomprince commented on src/_zkapauthorizer/storage_common.py:
```diff
@@ -43,27 +41,27 @@ class MorePassesRequired(Exception):
 
 def _message_maker(label):
     def make_message(storage_index):
-        return u"{label} {storage_index}".format(
+        return "{label} {storage_index}".format(
             label=label,
-            storage_index=b64encode(storage_index),
+            storage_index=b64encode(storage_index).decode("utf-8"),
         )
 
     return make_message
```
I guess it would be a wire-incompatible change, but I'd be inclined to change this to be `bytes`, rather that `str` here.

exarkun replied:
It turns out that all of the callers take the result and encode it to bytes - so I made this return bytes and simplified all of the callers.

tomprince commented on src/_zkapauthorizer/tests/test_storage_protocol.py:
```diff
@@ -1073,6 +1073,72 @@ def write():
+        )
+        self.assertThat(d, is_successful_write())
+
+        # Write some new data to with a correct test vector.  We can only be
```
s/to/too maybe?

tomprince reviewed (changes requested):
This looks good. I left a few comments and suggestions inline. Most of them are not blocking, but this is a significant enough change that I'd like to review any changes made.

I'll note that I only looked at the diff, and didn't try to find other places that needed to be changed. I also didn't changes of `b""` vs `u""` etc against the places being called, but mostly looked locally to verify that the changes appeared reasonable. I presume that anything significant will be caught by tests.

exarkun commented on src/_zkapauthorizer/foolscap.py:
```diff
@@ -90,7 +88,7 @@ def add_passes(schema):
         ``schema`` but with one additional required argument.
     """
-    return add_arguments(schema, [(b"passes", _PassList)])
+    return add_arguments(schema, [("passes", _PassList)])
```
Just so.

exarkun commented on src/_zkapauthorizer/controller.py:
```diff
@@ -78,7 +77,7 @@ class Unpaid(Exception):
 
 
-@attr.s(frozen=True)
+@attr.s
```
`auto_exc`, great, that is exactly the feature I needed to avoid removing frozen=True.  Also, auto_attributes: neat.

exarkun replied:
Oh no.  In fact it isn't. :(

The problem that `frozen=True` causes is with `Failure.cleanFailure`:

```
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/_zkapauthorizer/tests/test_controller.py", line 906, in test_redemption_unknown_response
    d = redeemer.redeemWithCounter(
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 1613, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 1529, in _cancellableInlineCallbacks
    _inlineCallbacks(None, g, status)
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 1464, in _inlineCallbacks
    status.deferred.errback()
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 501, in errback
    self._startRunCallbacks(fail)
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 568, in _startRunCallbacks
    self._runCallbacks()
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/internet/defer.py", line 699, in _runCallbacks
    current.result.cleanFailure()
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/twisted/python/failure.py", line 627, in cleanFailure
    self.value.__traceback__ = None
  File "/nix/store/bkyhzkq0wj55s9mvamv430jzy4ax9d70-python3-3.9.6-env/lib/python3.9/site-packages/attr/_make.py", line 626, in _frozen_setattrs
    raise FrozenInstanceError()
attr.exceptions.FrozenInstanceError
```

`Failure` tries to clear `__traceback__` and isn't allowed to. :/

exarkun replied:
And double oh-no!  `auto_exc` gives the instance identity-based comparison instead of value-based comparison and breaks expectations elsewhere.

meejah replied:
I concluded that with python3 twisted one can't use attrs `frozen=True` together with any exception classes.

tomprince replied:
We should maybe file a bug against twisted (or maybe attrs) about the traceback issue. The expectation issue could maybe be fixed by using `MatchesStructure`, but maybe that isn't worth it.

exarkun replied:
Looks like someone beat us to filing the bug(s):

https://twistedmatrix.com/trac/ticket/9641
https://twistedmatrix.com/trac/ticket/9771

I can add these links to our source.

exarkun commented on default.nix:
```diff
@@ -1,17 +1,19 @@
   sources = import nix/sources.nix;
 in
-{ pkgs ? import sources.release2105 {}
+{ pkgs ? import sources.release2111 { }
```
I think it will be easier to advance to 21.11 in PrivateStorageio than to back this up to 21.05.  There are some build failures in our dependencies on 21.05 that aren't interesting in any way once we're up to 21.11 and will be annoying to fix.  Since ZKAPAuthorizer is supplying Tahoe-LAFS, the fact that everything is working on 21.11 seems like a pretty strong signal that upgrading PrivateStorageio to 21.11 (at least for Tahoe-LAFS/ZKAPAuthorizer) will be straightforward.  Although it is true I haven't really tried it out yet.  I could do that before committing to this upgrade to confirm that is going to be the easier path.

exarkun commented on shell.nix:
```diff
@@ -3,12 +3,34 @@
-    tests.lint-python
-    pkgs.niv
+    # Provide the linting tools for interactive usage.
+    lint-python
```
I haven't yet internalized how black's opinions change over time and across different Python versions.

One thing to note is that `lint-python` pins Python 3.9 whereas `python-env` respects whatever value is passed in for `python` (currently defaulting to 3.9).

What's better?  Keeping the version of black as stable as possible (in that case, should we be pinning a black version in `tests.nix` too)?  Keeping black's Python runtime as stable as possible?  Making sure black's Python runtime matches our preferred deployment Python runtime?

tomprince replied:
Looking at black's FAQ, there is [this answer](https://black.readthedocs.io/en/stable/faq.html#how-stable-is-black-s-style), which suggests it is fairly stable. I think the only thing that makes a difference based on python version, is the target-version.

I was actually planning on suggesting change that in `pyproject.toml` (where it is currently set to `py27`). I think setting it to just `["py39"]` would be appropriate (though https://github.com/psf/black/issues/751 suggests we should maybe list all we want to support). This will perhaps require a new run of black (which will do things like removing `u""` prefixes).

To answer your question, I don't think pinning the python version matters. Theoretically, we should pin the black version, but I think in practice it isn't necessary.

exarkun replied:
Great.  Thanks for the links, especially.

Since switching the pyproject.toml black config will trigger a bunch of reformatting, I'd like to defer that to a follow-up branch (not long, just until after the Tahoe 1.17.1 PR is merged).

exarkun replied:
(https://github.com/PrivateStorageio/ZKAPAuthorizer/issues/283)

exarkun commented on src/_zkapauthorizer/resource.py:
```diff
@@ -401,14 +403,14 @@ def render_PUT(self, request):
-            return bad_request(u"json request body required").render(request)
-        if payload.keys() != [u"voucher"]:
+            return bad_request("json request body required").render(request)
+        if set(payload) != {"voucher"}:
```
I only recently learned that `dict_keys` will compare equal to a `set`.  It doesn't feel great to me.  I guess it's fine, though?  These codepaths seem to have good test coverage (100% line and branch coverage on this file!) so ... ok.

exarkun commented on src/_zkapauthorizer/storage_common.py:
```diff
@@ -148,9 +149,9 @@ def required_passes(bytes_per_pass, share_sizes):
     :return int: The number of passes required to cover the storage cost.
     """
-    if not isinstance(share_sizes, list):
+    if not isinstance(share_sizes, (list, _dict_values)):
```
`set` and `dict` are both iterable and bugs have been added at various points in the past having to do with accidentally or thoughtlessly using one of those types and having them reach this function so I think I don't want to open it all the way up to `Iterable`.  Maybe requiring exactly `list` would be better than having this `_dict_values` thing here but then all of the dict.values-using callers need to have `list(...)` added back...  Either way seems a little lame.  I'm inclined to leave it in this shape for now, I guess.

exarkun commented on src/_zkapauthorizer/tests/fixtures.py:
```diff
@@ -49,9 +47,9 @@ class AnonymousStorageServer(Fixture):
+        self.tempdir = FilePath(self.useFixture(TempDir()).join(u"storage"))
         self.storage_server = StorageServer(
-            self.tempdir.asBytesMode().path,
+            self.tempdir.asTextMode().path,
```
I'm not sure.  And I've let my local Windows dev environment fall into disrepair, I think.

On the other hand, it doesn't hurt much to be explicit about the type here right?  I know what I'm getting with `self.tempdir.asTextMode().path` but I'm not so sure if this were just `self.tempdir`.  It would depend on `FilePath` and `TempDir` and maybe also the `u"storage"` being passed in.

On Linux, looks like `TempDir.join` only accepts unicode (raises an exception when passed bytes) and returns unicode so I suppose the `asTextMode()` is superfluous here since the FilePath should already be wrapping a unicode path.

I think `StorageServer` will quickly blow up if we did manage to pass bytes here, since it calls `os.path.join` with the supplied value and a unicode string... So I guess removing the `asTextMode()` here is fine.

tomprince replied:
It looks like there is one remaining `asBytesMode()` call in `src/_zkapauthorizer/model.py`, which is passed to `sqlite3.connect`. I looked at the code in cpython for that (in both 3.9 and 3.10), and it supports bytes, strings and anything [path-like](https://docs.python.org/3/library/os.html#os.PathLike). I'd be inclined to remove that as well.

exarkun commented on src/_zkapauthorizer/controller.py:
```diff
@@ -340,24 +342,24 @@ class DummyRedeemer(object):
     the result.
 
-    :ivar unicode _public_key: The base64-encoded public key to return with
+    :ivar str _public_key: The base64-encoded public key to return with
```
It would be good to be consistent about the type used to represent the base64-encoded form of the various cryptographic values, yea.

exarkun replied:
#279

exarkun commented:
Thanks for the review @tomprince .  I think I've addressed most of the points.  For a few others, I've left some comments in reply to yours.  I think this is just about ready for a second look when you get a chance.  I see there are CI failures on macOS and Windows.  Since they don't happen for the Nix build I guess they have to do with version skew in some dependency and hopefully adding some tighter version pins will deal with them.

exarkun commented:
> Since they don't happen for the Nix build I guess they have to do with version skew in some dependency and hopefully adding some tighter version pins will deal with them.

Pretty sure this is about Tahoe-LAFS 1.17.1 which took RIStorageServer off of StorageServer.  So I changed the version pin to be exactly 1.17.0 for now.  This seems unrelated to the porting.  We didn't see it before because 1.17.1 was only released in the last day or two.

tomprince reviewed (approved):
This looks good. I think there are just three minor things to change (an `asBytesMode` call to remove, use `from __future__ import annotations` instead of `""` for type forward-reference, update the `target-version` for black and rerun).

Optionally, you can also remove the lint-python (in favour of just having one), but that could be a followup as well.

exarkun commented:
> Optionally, you can also remove the lint-python (in favour of just having one), but that could be a followup as well.

I started on this and quickly ran into problems where multiple conflicting instances of dependencies were showing up (autobahn, at least).  It wasn't immediately obvious why this was happening so I'll leave it for later.  I filed https://github.com/PrivateStorageio/ZKAPAuthorizer/issues/284
